guide/qip/qip-processor.rst

```diff
@@ -101,47 +181,222 @@ To let it find the optimal pulses, we need to give the parameters for :func:`qut
-The typical approaches are either applying bit/sign flipping gate probabilistically
-or applying Kraus operators representing different noisy channels (e.g. amplitude damping, dephasing) after each unitary gate evolution. In the case of single qubit, they have the same effect and the parameters in the Kraus operators are exactly the probability of a flipping error happens during the gate operation time.
+.. note::
 
-Since the processor simulates the state evolution at the level of the driving Hamiltonian, there is no way to apply an error operator to the continuous time evolution. Instead, the error is added to the driving Hamiltonian list (coherent control error) or the collapse operators (decoherent error) contributing to the evolution. Mathematically, this is no different from adding error channel probabilistically (it is actually how :func:`qutip.mcsolve` works internally). The collapse operator for single-qubit amplitude damping and dephasing are exactly the destroying operator and the sign-flipping operator. One just needs to choose the correct coefficients for them to simulate the noise, e.g. the relaxation time T1 and dephasing time T2. Because it is based on the open system evolution instead of abstract operators, this simulation is closer to the physical implementation and requires less pre-analysis of the system.
+   Available from QuTiP 4.6
 
-Compared to the approach of Kraus operators, this way of simulating noise is more computationally expensive. If you only want to simulate the decoherence of single-qubit relaxation, there is no need to go through all the calculations. However, this simulator is closer to the real experiment and, therefore, more convenient in some cases, such as when coherent noise or correlated noise exist. For instance, a pulse on one qubit might affect the neighbouring qubits, the evolution is still unitary but the gate fidelity will decrease. It is not always easy or even possible to define a noisy gate matrix. In our simulator, it can be done by defining a :class:`qutip.qip.device.ControlAmpNoise` (Control Amplitude Noise). Here we show two examples (the source code can be found in the gallery):
+In order to simulate quantum circuits at the level of time evolution.
+We need to first compile the circuit into the Hamiltonian model, i.e.
+the control pulses.
+Hence each :class:`~qutip.qip.device.Processor` has a corresponding 
+:class:`~qutip.qip.compiler.GateCompiler` class.
+The compiler takes a :class:`~qutip.qip.circuit.QubitCircuit`
+and returns the compiled ``tlist`` and ``coeffs``.
+It is called implicitly when calling the method
+:class:`~qutip.qip.device.Processor.run_state`.
 
-The first example is a processor with one qubit under rotation around the z-axis and relaxation time :math:`T_2=5`. We measure the population of the :math:`\left| + \right\rangle` state and observe the Ramsey signal:
+.. testcode::
 
-.. image:: /gallery/qutip_examples/qip/images/sphx_glr_plot_qip_relaxation_001.png
```

Comment:
if these images aren't used any more, do the files need to be removed from the repo?

Reply:
They are still in the gallery. I just find it's better to also show the code here. The gallery is not used by many...

Those ones in the gallery are generated automatically at the run-time.